File: internal/assets/plugins/cortex-snapshot.ts
// OpenCode Plugin helper ensuring default export is a valid plugin definition object for v1 and v2
export const Plugin = {
  define: <T extends { id: string; setup?: (ctx: any) => Promise<any> | any; server?: (ctx: any) => Promise<any> | any }>(def: T): T => {
    if (!def.server && def.setup) {
      def.server = def.setup;
    }
    return def;
  },
};

import { Buffer } from "node:buffer";
import { execFileSync } from "node:child_process";
import { createHash } from "node:crypto";

export const CortexSnapshotPlugin = Plugin.define({
  id: "cortex-snapshot",
  async setup(ctx: any) {
    const toolDef = {
      name: "cortex_ia_snapshot_read",
      description: "Read one exact local Cortex observation through Cortex-IA's bounded streaming project export. Verifies identity and optional SHA-256; returns content and digest. The upstream still exports the project (64 MiB total, 2 MiB per record, 1 MiB selected content, 30 seconds); no project array is retained. Never substitutes for a remote MCP store.",
      input: {
        type: "object",
        properties: {
          observation_id: { type: "integer", description: "Positive safe integer observation ID" },
          project: { type: "string", description: "Project identifier (1-512 UTF-8 bytes)" },
          expected_sha256: { type: "string", description: "Optional lowercase SHA-256 digest" },
        },
        required: ["observation_id", "project"],
        additionalProperties: false,
      },
      async execute(args: any) {
        if (!Number.isSafeInteger(args.observation_id) || args.observation_id < 1) throw new Error("observation_id must be a positive safe integer");
        if (!args.project?.trim() || Buffer.byteLength(args.project, "utf8") > 512) throw new Error("project must contain 1-512 UTF-8 bytes");
        if (args.expected_sha256 !== undefined && !/^[a-f0-9]{64}$/.test(args.expected_sha256)) throw new Error("expected_sha256 must be a lowercase SHA-256 digest");
        const command = ["snapshot", "read", "--project", args.project, "--id", String(args.observation_id)];
        if (args.expected_sha256 !== undefined) command.push("--expected-sha256", args.expected_sha256);
        let raw: string;
        try {
          raw = execFileSync("cortex-ia", command, { encoding: "utf8", maxBuffer: 8 * 1024 * 1024, timeout: 35000, windowsHide: true });
        } catch {
          throw new Error("Local Cortex snapshot failed, timed out or exceeded bounds; no verified snapshot available");
        }
        let snapshot: any;
        try { snapshot = JSON.parse(raw); } catch { throw new Error("Invalid structured Cortex snapshot"); }
        if (snapshot?.transport !== "local_cortex_cli" || snapshot.project !== args.project || snapshot.observation_id !== args.observation_id || typeof snapshot.content !== "string") throw new Error("Snapshot identity mismatch");
        const bytes = Buffer.from(snapshot.content, "utf8");
        if (bytes.length > 1048576 || bytes.toString("utf8") !== snapshot.content || snapshot.byte_length !== bytes.length) throw new Error("Invalid snapshot content or byte length");
        const digest = createHash("sha256").update(bytes).digest("hex");
        if (snapshot.sha256 !== digest || (args.expected_sha256 !== undefined && args.expected_sha256 !== digest)) throw new Error("Snapshot digest mismatch");
        return { content: JSON.stringify(snapshot) };
      }
    };

    if (ctx?.tool?.transform) {
      await ctx.tool.transform((editor: any) => {
        editor.add(toolDef);
      });
    }

    const cleanup = async () => {};
    (cleanup as any).dispose = cleanup;
    (cleanup as any).tool = {
      cortex_ia_snapshot_read: toolDef,
    };
    return cleanup;
  }
});

export default CortexSnapshotPlugin;

File: internal/delegation/cortex_snapshot.go
package delegation

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"time"
	"unicode/utf8"
)

const (
	cortexExportLimit  = 64 * 1024 * 1024
	cortexRecordLimit  = 2 * 1024 * 1024
	cortexContentLimit = 1024 * 1024
)

type CortexSnapshotRequest struct {
	ID             int64
	ExpectedSHA256 string
}

type CortexSnapshot struct {
	Transport     string `json:"transport"`
	Project       string `json:"project"`
	ObservationID int64  `json:"observation_id"`
	Content       string `json:"content"`
	SHA256        string `json:"sha256"`
	ByteLength    int    `json:"byte_length"`
}

// ReadCortexSnapshots streams the existing project export, not a per-ID query.
// Success requires the complete export AND successful producer termination.
func ReadCortexSnapshots(ctx context.Context, project string, requests []CortexSnapshotRequest) ([]CortexSnapshot, error) {
	if err := validateSnapshotRequests(project, requests); err != nil {
		return nil, err
	}
	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
	defer cancel()
	command := exec.CommandContext(ctx, "cortex", "export", "--project", project)
	command.WaitDelay = time.Second
	return readCortexSnapshotCommand(command, cancel, project, requests)
}

func validateSnapshotRequests(project string, requests []CortexSnapshotRequest) error {
	if strings.TrimSpace(project) == "" || len(project) > 512 || !utf8.ValidString(project) || len(requests) == 0 || len(requests) > 128 {
		return errors.New("snapshot requires a valid project and 1..128 observations")
	}
	seen := make(map[int64]bool)
	for _, request := range requests {
		if request.ID < 1 || request.ID > 9007199254740991 || seen[request.ID] {
			return errors.New("snapshot IDs must be unique positive safe integers")
		}
		seen[request.ID] = true
		if request.ExpectedSHA256 != "" {
			b, err := hex.DecodeString(request.ExpectedSHA256)
			if err != nil || len(b) != sha256.Size || strings.ToLower(request.ExpectedSHA256) != request.ExpectedSHA256 {
				return errors.New("snapshot digest must be lowercase SHA-256")
			}
		}
	}
	return nil
}

func readCortexSnapshotCommand(command *exec.Cmd, cancel context.CancelFunc, project string, requests []CortexSnapshotRequest) ([]CortexSnapshot, error) {
	stdout, err := command.StdoutPipe()
	if err != nil {
		return nil, err
	}
	// Stderr is drained but never retained or reflected into errors (it may contain private data).
	stderr := &cortexStderr{cancel: cancel}
	command.Stderr = stderr
	if err = command.Start(); err != nil {
		return nil, errors.New("local Cortex export could not start")
	}
	snapshots, readErr := decodeCortexSnapshots(stdout, project, requests)
	if readErr != nil {
		cancel()
		_ = stdout.Close()
	}
	waitErr := command.Wait()
	if readErr != nil {
		return nil, readErr
	}
	if waitErr != nil || stderr.bytes > 64*1024 {
		return nil, errors.New("local Cortex export failed, timed out or exceeded stderr bounds; snapshot remains unverified")
	}
	return snapshots, nil
}

type cortexStderr struct {
	bytes  int
	cancel context.CancelFunc
}

func (s *cortexStderr) Write(p []byte) (int, error) {
	s.bytes += len(p)
	if s.bytes > 64*1024 {
		s.cancel()
		return 0, errors.New("local Cortex stderr limit exceeded")
	}
	return len(p), nil
}

var jsonEscapePattern = regexp.MustCompile(`\\(?:u[0-9a-fA-F]{4}|.)`)

// encoding/json replaces lone UTF-16 surrogates with U+FFFD. Reject that lossy
// normalization before hashing, while allowing literal U+FFFD and valid pairs.
func exactJSONUnicode(raw []byte) bool {
	escapes := jsonEscapePattern.FindAllIndex(raw, -1)
	for i := 0; i < len(escapes); i++ {
		start, end := escapes[i][0], escapes[i][1]
		if end-start != 6 || raw[start+1] != 'u' {
			continue
		}
		value, _ := strconv.ParseUint(string(raw[start+2:end]), 16, 16)
		if value >= 0xdc00 && value <= 0xdfff {
			return false
		}
		if value >= 0xd800 && value <= 0xdbff {
			if i+1 >= len(escapes) || escapes[i+1][0] != end || escapes[i+1][1]-end != 6 || raw[end+1] != 'u' {
				return false
			}
			low, _ := strconv.ParseUint(string(raw[end+2:end+6]), 16, 16)
			if low < 0xdc00 || low > 0xdfff {
				return false
			}
			i++
		}
	}
	return true
}

// Bound decoder read-ahead as well as each record, so a giant unselected record
// cannot force a project-sized allocation before Decode returns.
type cortexExportReader struct {
	source    io.Reader
	read, end int64
}

func (r *cortexExportReader) Read(p []byte) (int, error) {
	remaining := min(int64(cortexExportLimit), r.end) - r.read
	if remaining <= 0 {
		return 0, errors.New("local Cortex export exceeds total or record byte limit")
	}
	if int64(len(p)) > remaining {
		p = p[:remaining]
	}
	n, err := r.source.Read(p)
	r.read += int64(n)
	return n, err
}

func decodeCortexSnapshots(source io.Reader, project string, requests []CortexSnapshotRequest) ([]CortexSnapshot, error) {
	if err := validateSnapshotRequests(project, requests); err != nil {
		return nil, err
	}
	input := &cortexExportReader{source: source, end: cortexRecordLimit}
	decoder := json.NewDecoder(input)
	token, err := decoder.Token()
	if err != nil || token != json.Delim('[') {
		return nil, errors.New("local Cortex export must be a JSON array")
	}
	wanted := make(map[int64]string, len(requests))
	for _, request := range requests {
		wanted[request.ID] = request.ExpectedSHA256
	}
	found := make(map[int64]CortexSnapshot, len(requests))
	for {
		input.end = decoder.InputOffset() + cortexRecordLimit
		if !decoder.More() {
			break
		}
		var raw json.RawMessage
		if err := decoder.Decode(&raw); err != nil {
			return nil, errors.New("local Cortex export has malformed, truncated or oversized records")
		}
		if len(raw) > cortexRecordLimit || !utf8.Valid(raw) || !exactJSONUnicode(raw) {
			return nil, errors.New("local Cortex export record exceeds bounds or has invalid Unicode")
		}
		var observation struct {
			ID      int64   `json:"id"`
			Project *string `json:"project"`
			Content *string `json:"content"`
		}
		if err := json.Unmarshal(raw, &observation); err != nil || observation.ID < 1 {
			return nil, errors.New("local Cortex export has an invalid observation identity")
		}
		expected, selected := wanted[observation.ID]
		if !selected {
			continue
		}
		if _, duplicate := found[observation.ID]; duplicate {
			return nil, errors.New("local Cortex export contains a duplicate requested observation")
		}
		if observation.Project == nil || *observation.Project != project || observation.Content == nil || len(*observation.Content) > cortexContentLimit {
			return nil, errors.New("local Cortex snapshot project or content does not match requested identity and bounds")
		}
		digest := sha256.Sum256([]byte(*observation.Content))
		encoded := hex.EncodeToString(digest[:])
		if expected != "" && expected != encoded {
			return nil, errors.New("local Cortex snapshot digest mismatch; contract validation failed")
		}
		found[observation.ID] = CortexSnapshot{"local_cortex_cli", project, observation.ID, *observation.Content, encoded, len(*observation.Content)}
	}
	if token, err = decoder.Token(); err != nil || token != json.Delim(']') {
		return nil, errors.New("local Cortex export array is truncated")
	}
	input.end = decoder.InputOffset() + cortexRecordLimit
	if err = decoder.Decode(new(json.RawMessage)); err != io.EOF {
		return nil, errors.New("local Cortex export has trailing data or exceeds bounds")
	}
	results := make([]CortexSnapshot, 0, len(requests))
	for _, request := range requests {
		snapshot, exists := found[request.ID]
		if !exists {
			return nil, fmt.Errorf("local Cortex observation %s missing from export", strconv.FormatInt(request.ID, 10))
		}
		results = append(results, snapshot)
	}
	return results, nil
}
